File: viewer-config-persistence/src/main/java/nl/b3p/viewer/config/services/FeatureTypeRelation.java
/*
 * Copyright (C) 2013 B3Partners B.V.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package nl.b3p.viewer.config.services;

import java.util.ArrayList;
import java.util.List;
import javax.persistence.CascadeType;
import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 *
 * @author Roy Braam
 */
@Entity
public class FeatureTypeRelation {
    public static final String JOIN = "join";
    public static final String RELATE = "relate";
    
    @Id
    private Long id;
    @ManyToOne    
    private SimpleFeatureType featureType;
    
    @ManyToOne
    private SimpleFeatureType foreignFeatureType;
        
    @OneToMany(cascade=CascadeType.ALL, mappedBy="relation", orphanRemoval=true)
    private List<FeatureTypeRelationKey> relationKeys = new ArrayList<FeatureTypeRelationKey>();
    
    private String type;
    

    //<editor-fold defaultstate="collapsed" desc="Getters/Setters">
    public Long getId() {
        return id;
    }
    
    public void setId(Long id) {
        this.id = id;
    }
    
    public SimpleFeatureType getFeatureType() {
        return featureType;
    }
    
    public void setFeatureType(SimpleFeatureType featureType) {
        this.featureType = featureType;
    }
    
    public List<FeatureTypeRelationKey> getRelationKeys() {
        return relationKeys;
    }
    
    public void setRelationKeys(List<FeatureTypeRelationKey> relationKeys) {
        this.relationKeys = relationKeys;
    }
    
    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public SimpleFeatureType getForeignFeatureType() {
        return foreignFeatureType;
    }

    public void setForeignFeatureType(SimpleFeatureType foreignFeatureType) {
        this.foreignFeatureType = foreignFeatureType;
    }
    //</editor-fold>

    public JSONObject toJSONObject() throws JSONException {
        JSONObject j = new JSONObject();
        j.put("type",type); 
        if (this.featureType!=null){
            j.put("featureType", this.featureType.getId());
        }
        if (this.foreignFeatureType!=null){
            j.put("foreignFeatureType", this.foreignFeatureType.getId());
            JSONArray jRel = new JSONArray();
            if (!this.foreignFeatureType.getRelations().isEmpty()){
                j.put("relations",jRel);
                for (FeatureTypeRelation rel : this.foreignFeatureType.getRelations()){
                    jRel.put(rel.toJSONObject());
                }
            }
        }
        return j;
    }

}

File: viewer-admin/src/main/webapp/resources/js/services/featuretyperelation.js
/*
 * Copyright (C) 2012-2016 B3Partners B.V.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
Ext.define('vieweradmin.components.FeaturetypeRelation', {

    extend: "Ext.ux.b3p.CrudGrid",

    config: {
        gridurl: "",
        editurl: "",
        deleteurl: "",
        itemname: "relaties",
        editattributesurl: ""
    },

    constructor: function(config) {
        this.initConfig(config);
        vieweradmin.components.FeaturetypeRelation.superclass.constructor.call(this, this.config);
        vieweradmin.components.Menu.setActiveLink('menu_relation');
    },

    getGridColumns: function() {
        return [
            {
                id: 'featuretype',
                text: i18next.t('viewer_admin_featuretyperelation_0'),
                dataIndex: 'featuretype',
                flex: 1,
                filter: {
                    xtype: 'textfield'
                }
            },{
                id: 'foreignFeaturetype',
                text: i18next.t('viewer_admin_featuretyperelation_1'),
                dataIndex: 'foreignFeaturetype',
                flex: 1,
                filter: {
                    xtype: 'textfield'
                }
            },{
                id: 'edit',
                header: '',
                dataIndex: 'id',
                width: 200,
                sortable: false,
                hideable: false,
                menuDisabled: true,
                renderer: (function(value) {
                    return [
                        Ext.String.format('<a href="#" class="editobject">' + ___("Bewerken") + '</a>'),
                        Ext.String.format('<a href="#" class="removeobject">' + ___("Verwijderen") + '</a>')
                    ].join(" | ");
                }).bind(this)
            }
        ];
    },

    getGridModel: function() {
        return [
            {name: 'id', type: 'int' },
            {name: 'featuretype', type: 'string'},
            {name: 'foreignFeaturetype', type: 'string'}
        ];
    },

    removeConfirmMessage: function(record) {
        return ["Weet u zeker dat u deze relatie wilt verwijderen?"].join("");
    },

    getEditUrl: function(record) {
        return this.createUrl(this.config.editurl, { relation: record.get('id') });
    },

    getRemoveUrl: function(record) {
        return this.createUrl(this.config.deleteurl, { relation: record.get('id') });
    }

});